File: roll-your-own-mock-framework/typescript/src/Mock.ts
import { CallRecord } from './CallRecord.js';
import { VerificationError } from './VerificationError.js';

interface StubEntry {
  args: unknown[];
  returnValue: unknown;
}

interface MethodVerification {
  wasCalled(): void;
  wasCalledWith(...args: unknown[]): void;
  wasCalledTimes(n: number): void;
}

interface StubConfiguration {
  thenReturn(value: unknown): void;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type MockProxy = Record<string, (...args: any[]) => unknown>;

class MockState {
  readonly calls: CallRecord[] = [];
  private readonly stubs: Map<string, StubEntry[]> = new Map();

  recordCall(methodName: string, args: unknown[]): void {
    this.calls.push({ methodName, args });
  }

  addStub(methodName: string, args: unknown[], returnValue: unknown): void {
    if (!this.stubs.has(methodName)) {
      this.stubs.set(methodName, []);
    }
    this.stubs.get(methodName)!.push({ args, returnValue });
  }

  findStub(methodName: string, args: unknown[]): unknown {
    const entries = this.stubs.get(methodName);
    if (!entries) return undefined;
    for (const entry of entries) {
      if (argsMatch(entry.args, args)) {
        return entry.returnValue;
      }
    }
    return undefined;
  }
}

function argsMatch(expected: unknown[], actual: unknown[]): boolean {
  if (expected.length !== actual.length) return false;
  return expected.every((val, i) => val === actual[i]);
}

function formatArgs(args: unknown[]): string {
  return args.map(a => String(a)).join(', ');
}

const stateMap = new WeakMap<MockProxy, MockState>();

export function createMock(): MockProxy {
  const state = new MockState();

  const proxy = new Proxy({} as MockProxy, {
    get(_target, prop: string) {
      return (...args: unknown[]): unknown => {
        state.recordCall(prop, args);
        return state.findStub(prop, args);
      };
    },
  });

  stateMap.set(proxy, state);
  return proxy;
}

export function when(mock: MockProxy): MockProxy {
  const state = stateMap.get(mock);
  if (!state) throw new Error('not a mock');

  return new Proxy({} as MockProxy, {
    get(_target, prop: string) {
      return (...args: unknown[]): StubConfiguration => ({
        thenReturn(value: unknown): void {
          state.addStub(prop, args, value);
        },
      });
    },
  });
}

export function verify(mock: MockProxy): Record<string, MethodVerification> {
  const state = stateMap.get(mock);
  if (!state) throw new Error('not a mock');

  return new Proxy({} as Record<string, MethodVerification>, {
    get(_target, prop: string): MethodVerification {
      return {
        wasCalled(): void {
          const called = state.calls.some(c => c.methodName === prop);
          if (!called) {
            throw new VerificationError(`expected ${prop} to be called but was never called`);
          }
        },

        wasCalledWith(...expectedArgs: unknown[]): void {
          const methodCalls = state.calls.filter(c => c.methodName === prop);
          if (methodCalls.length === 0) {
            throw new VerificationError(
              `expected ${prop}(${formatArgs(expectedArgs)}) to be called but was never called`,
            );
          }
          const match = methodCalls.some(c => argsMatch(expectedArgs, c.args));
          if (!match) {
            throw new VerificationError(
              `expected ${prop}(${formatArgs(expectedArgs)}) to be called but was called with (${formatArgs(methodCalls[0]!.args)})`,
            );
          }
        },

        wasCalledTimes(n: number): void {
          const actualCount = state.calls.filter(c => c.methodName === prop).length;
          if (actualCount !== n) {
            throw new VerificationError(
              `expected ${prop} to be called ${n} times but was called ${actualCount} times`,
            );
          }
        },
      };
    },
  });
}

File: roll-your-own-mock-framework/python/src/roll_your_own_mock_framework/mock.py
from typing import Any

from .call_record import CallRecord
from .verification_error import VerificationError


def _format_args(args: tuple[Any, ...]) -> str:
    return ", ".join(str(a) for a in args)


def _args_match(expected: tuple[Any, ...], actual: tuple[Any, ...]) -> bool:
    return expected == actual


class _StubConfiguration:
    def __init__(self, mock: "Mock", method_name: str, args: tuple[Any, ...]) -> None:
        self._mock = mock
        self._method_name = method_name
        self._args = args

    def then_return(self, value: Any) -> None:
        self._mock._add_stub(self._method_name, self._args, value)


class _MethodVerification:
    def __init__(self, mock: "Mock", method_name: str) -> None:
        self._mock = mock
        self._method_name = method_name

    def was_called(self) -> None:
        called = any(c.method_name == self._method_name for c in self._mock.calls)
        if not called:
            raise VerificationError(
                f"expected {self._method_name} to be called but was never called"
            )

    def was_called_with(self, *expected_args: Any) -> None:
        method_calls = [
            c for c in self._mock.calls if c.method_name == self._method_name
        ]
        if not method_calls:
            raise VerificationError(
                f"expected {self._method_name}({_format_args(expected_args)}) to be called but was never called"
            )
        match = any(_args_match(expected_args, c.args) for c in method_calls)
        if not match:
            actual_args = method_calls[0].args
            raise VerificationError(
                f"expected {self._method_name}({_format_args(expected_args)}) to be called but was called with ({_format_args(actual_args)})"
            )

    def was_called_times(self, n: int) -> None:
        actual_count = sum(
            1 for c in self._mock.calls if c.method_name == self._method_name
        )
        if actual_count != n:
            raise VerificationError(
                f"expected {self._method_name} to be called {n} times but was called {actual_count} times"
            )


class _WhenProxy:
    def __init__(self, mock: "Mock") -> None:
        self._mock = mock

    def __getattr__(self, name: str) -> Any:
        def capture_args(*args: Any) -> _StubConfiguration:
            return _StubConfiguration(self._mock, name, args)

        return capture_args


class _VerifyProxy:
    def __init__(self, mock: "Mock") -> None:
        self._mock = mock

    def __getattr__(self, name: str) -> _MethodVerification:
        return _MethodVerification(self._mock, name)


class Mock:
    def __init__(self) -> None:
        self._calls: list[CallRecord] = []
        self._stubs: dict[str, list[tuple[tuple[Any, ...], Any]]] = {}

    @property
    def calls(self) -> list[CallRecord]:
        return self._calls

    def __getattr__(self, name: str) -> Any:
        if name.startswith("_"):
            raise AttributeError(name)

        def method_proxy(*args: Any) -> Any:
            self._calls.append(CallRecord(method_name=name, args=args))
            stubs = self._stubs.get(name, [])
            for stub_args, return_value in stubs:
                if _args_match(stub_args, args):
                    return return_value
            return None

        return method_proxy

    def _add_stub(
        self, method_name: str, args: tuple[Any, ...], return_value: Any
    ) -> None:
        if method_name not in self._stubs:
            self._stubs[method_name] = []
        self._stubs[method_name].append((args, return_value))


def create_mock() -> Mock:
    return Mock()


def when(mock: Mock) -> _WhenProxy:
    return _WhenProxy(mock)


def verify(mock: Mock) -> _VerifyProxy:
    return _VerifyProxy(mock)
